jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PushLayerStep.java

```diff
@@ -0,0 +1,91 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.builder.steps;
+
+import com.google.cloud.tools.jib.api.RegistryException;
+import com.google.cloud.tools.jib.blob.BlobDescriptor;
+import com.google.cloud.tools.jib.builder.ProgressEventDispatcher;
+import com.google.cloud.tools.jib.builder.TimerEventDispatcher;
+import com.google.cloud.tools.jib.cache.CachedLayer;
+import com.google.cloud.tools.jib.configuration.BuildConfiguration;
+import com.google.cloud.tools.jib.http.Authorization;
+import com.google.common.collect.ImmutableList;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+class PushLayerStep implements Callable<BlobDescriptor> {
+
+  private static final String DESCRIPTION = "Preparing application layer pushers";
+
+  private final BuildConfiguration buildConfiguration;
+  private final ProgressEventDispatcher.Factory progressEventDispatcherFactory;
+
+  private final Authorization pushAuthorization;
+  private final Future<CachedLayerAndName> cachedLayerAndName;
+
+  static ImmutableList<PushLayerStep> makeList(
+      BuildConfiguration buildConfiguration,
+      ProgressEventDispatcher.Factory progressEventDispatcherFactory,
+      Authorization pushAuthorization,
+      List<Future<CachedLayerAndName>> cachedLayers) {
+    try (TimerEventDispatcher ignored =
+            new TimerEventDispatcher(buildConfiguration.getEventHandlers(), DESCRIPTION);
+        ProgressEventDispatcher progressEventDispatcher =
+            progressEventDispatcherFactory.create(
+                "Preparing application layer pushers", cachedLayers.size())) {
+
+      // Constructs a PushBlobStep for each layer.
+      List<PushLayerStep> blobPushers = new ArrayList<>();
+      for (Future<CachedLayerAndName> layer : cachedLayers) {
+        ProgressEventDispatcher.Factory childProgressProducer =
+            progressEventDispatcher.newChildProducer();
+        blobPushers.add(
+            new PushLayerStep(buildConfiguration, childProgressProducer, pushAuthorization, layer));
+      }
+
+      return ImmutableList.copyOf(blobPushers);
+    }
+  }
+
+  PushLayerStep(
+      BuildConfiguration buildConfiguration,
+      ProgressEventDispatcher.Factory progressEventDispatcherFactory,
+      Authorization pushAuthorization,
+      Future<CachedLayerAndName> cachedLayerAndName) {
+    this.buildConfiguration = buildConfiguration;
+    this.progressEventDispatcherFactory = progressEventDispatcherFactory;
+    this.pushAuthorization = pushAuthorization;
+    this.cachedLayerAndName = cachedLayerAndName;
+  }
+
+  @Override
+  public BlobDescriptor call()
+      throws IOException, RegistryException, ExecutionException, InterruptedException {
+    CachedLayer layer = cachedLayerAndName.get().getCachedLayer();
```

Comment:
Note to reviewers: this is the blocking point for the step, and the only place for blocking in any step among all.

Reply:
I'm not sure I understand the scope of this statement? Is this the only place we call `get()` and expect it to wait?